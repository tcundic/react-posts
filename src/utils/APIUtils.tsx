import {apiEndpoint} from "../config";
import RequestOptions from "../interfaces/Request";
import Post from "../interfaces/Post";

const request = (options: RequestOptions, additionalHeaders?: Object): Promise<any> => {
    const headers = {
        'Content-Type': 'application/json',
        ...additionalHeaders
    };

    const defaults = {headers};
    options = Object.assign({}, defaults, options, {crossDomain: true});

    return fetch(options.url, options)
        .then(response => response.json());
};

export function getPosts(): Promise<Post[]> {
    return request({
        url: `${apiEndpoint}posts`,
        method: 'GET'
    });
}

export function getPostById(id: number): Promise<Post> {
    return request({
        url: `${apiEndpoint}posts/${id}`,
        method: 'GET'
    });
}
