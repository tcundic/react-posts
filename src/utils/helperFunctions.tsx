export const greet = (greeting: string, componentName: string): void => console.log(`${greeting} ${componentName}`);