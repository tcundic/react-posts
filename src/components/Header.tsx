import React, {useEffect, useState} from 'react';
import {greet} from "../utils/helperFunctions";

const Header = (
    {greeting, navItems, logo, title}:
        {
            greeting: string,
            navItems: [{ title: string, url: string }],
            logo: string,
            title: string
        }) => {

    useEffect(() => greet(greeting, Header.name));

    const [isMenuOpened, setIsMenuOpened] = useState(false);

    return (
        <header>
            <div className="container">
                <nav className="navbar">
                    <div className="navbar-brand mr-6">
                        <a className="navbar-item" href="/">
                            <img src={logo} className="app-logo" alt="react-logo"/>
                            {title}
                        </a>
                        <div className="navbar-burger" data-target="navbar-menu" onClick={() => setIsMenuOpened(prevState => !prevState)}>
                            <span></span>
                            <span></span>
                            <span></span>
                        </div>
                    </div>
                    <div id="navbar-menu" className={`navbar-menu ${isMenuOpened ? 'is-active' : ''}`}>
                        <div className="navbar-end">
                            {navItems.map(item =>
                                <a className="navbar-item" href={item.url} key={item.url}>
                                    {item.title}
                                </a>
                            )}
                        </div>
                    </div>
                </nav>
            </div>
        </header>
    );
};

export default Header;
