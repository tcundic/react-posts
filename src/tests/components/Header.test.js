import React from 'react';
import Header from "../../components/Header";
import { mount, shallow } from 'enzyme';
import { expect as chaiExpect } from "chai";
import toJson from "enzyme-to-json";
import navItems from "../fixtures/routes";
import * as Greet from "../../utils/helperFunctions";
import { BrowserRouter } from "react-router-dom";

const greeting = "Hello from";
const title = "React app";
const logo = "./logo";
const name = Header.name;

test('should render Header correctly', () => {
    const wrapper = shallow(<Header greeting={greeting} title={title} logo={logo} navItems={navItems} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

test('should print Header greeting', () => {
    const greetSpy = jest.spyOn(Greet, 'greet');
    const wrapper = mount(<BrowserRouter><Header greeting={greeting} title={title} logo={logo} navItems={navItems} /></BrowserRouter>);
    expect(greetSpy).toHaveBeenLastCalledWith(greeting, name);
});

test('should render 2 nav item', () => {
    const wrapper = shallow(<Header greeting={greeting} title={title} logo={logo} navItems={navItems} />);
    chaiExpect(wrapper.find('#navbar-menu').children().find('.navbar-item')).to.have.length(2);
});
