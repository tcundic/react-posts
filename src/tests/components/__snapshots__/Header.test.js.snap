// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <div
    className="container"
  >
    <nav
      className="navbar"
    >
      <div
        className="navbar-brand mr-6"
      >
        <NavLink
          className="navbar-item"
          to="/"
        >
          <img
            alt="react-logo"
            className="app-logo"
            src="./logo"
          />
          React app
        </NavLink>
        <div
          className="navbar-burger"
          data-target="navbar-menu"
          onClick={[Function]}
        >
          <span />
          <span />
          <span />
        </div>
      </div>
      <div
        className="navbar-menu "
        id="navbar-menu"
      >
        <div
          className="navbar-end"
        >
          <NavLink
            activeClassName="active-item"
            className="navbar-item"
            key="/item1"
            to="/item1"
          >
            Item1
          </NavLink>
          <NavLink
            activeClassName="active-item"
            className="navbar-item"
            key="/item2"
            to="/item2"
          >
            Item2
          </NavLink>
        </div>
      </div>
    </nav>
  </div>
</header>
`;
